text/204-golang-bindings.md

```diff
@@ -0,0 +1,669 @@
+would need to generate an interface in addition to a concrete struct for each JSII class.
+
+The JSII [ClassType](https://github.com/aws/jsii/blob/master/packages/%40jsii/spec/lib/assembly.ts#L803) provides information on whether a
+class is abstract, whether it extends another class, and whether it implements other interfaces. We will discuss each case in turn.
+
+### Case 1: Simple class
+
+Example (Taken from [Typescript handbook](https://www.typescriptlang.org/docs/handbook/classes.html#classes)):
+
+```ts
+class Greeter {
+    readonly greeting: string;
+
+    constructor(message: string) {
+        this.greeting = message;
+    }
+
+    greet(): string {
+        return "Hello, " + this.greeting;
+    }
+}
+
+let greeter = new Greeter("world");
+greeter.greet() // "Hello, world"
+```
+
+In Go: ([Go playground example](https://play.golang.org/p/cMW3Rv0lERh))
+
+```go
+package greeter
+
+type GreeterIface interface {
+    Greet() string
+}
+
+type Greeter struct {
+    greeting string
+}
+
+func New(message string) *Greeter {
+    return &Greeter{message}
+}
+
+func (g *Greeter) Greet() string {
+    return fmt.Sprintf("Hello, %+s", g.greeting)
+}
+
+// usage
+g := greeter.New("world")
+fmt.Println(g.Greet()) // "Hello, world"
+```
+
+### Case 2: Extending a Base class
+
+Example (taken from [here](https://www.typescriptlang.org/docs/handbook/classes.html#inheritance)):
+
+```ts
+class Animal {
+    public readonly name: string;
+
+    constructor(theName: string) { this.name = theName; }
+
+    public move(distanceInMeters: number = 0) {
+        console.log(`${this.name} moved ${distanceInMeters}m.`);
+    }
+}
+
+class Snake extends Animal {
+    constructor(name: string) { super(name); }
+    public move(distanceInMeters = 5) {
+        console.log("Slithering...");
+        super.move(distanceInMeters);
+    }
+}
+
+const sam = new Snake("Sammy the Python");
+
+sam.move();
+//  Slithering...
+// Sammy the Python moved 10m.
+```
+
+In Go:
+([Go playground example](https://play.golang.org/p/ZwoKmLn5OkV))
+([Example with embedded field as named property](https://play.golang.org/p/nWfmHhxbs1G))
+
+```go
+package animal
+
+// Base class as interface
+type AnimalIface interface {
+    Name() string
+    Move(distance int64)
+}
+
+// Base class implementation
+type Animal struct {
+    name string
+}
+
+func NewAnimal(name string) AnimalIface {
+    return &Animal{name}
+}
+
+func (a *Animal) Name() string {
+    return a.name
+}
+
+func (a *Animal) Move(distance int64) {
+    fmt.Printf("%s moved %vm.\n", a.Name(), distance)
+}
+
+// Child class
+type SnakeIface interface {
+  AnimalIface
+}
+
+type Snake struct {
+    Animal
+}
+
+func NewSnake(name string) SnakeIface {
+    return &Snake{Animal{name}}
+}
+
+func (s *Snake) Name() string {
+    return s.Animal.Name()      // simulates "super" call
```

Comment:
I thought we were supposed to get automatic delegation in Go. Is that not true, do we manually need to forward every inherited method?

Reply:
Yes, this can be implemented:
```go
func (s *Snake) Name() string {
	return s.name
}
```
Run: https://play.golang.org/p/v9gVW0bG2y1

I think I was trying to make a point about the super call but perhaps that makes it more confusing.